File: Other Algorithms/KMP/java/KMP.java
public class KMP {

	public static void main(String[] args) {

		String text="piyushpiy";
		String pat="piy";

		int n = text.length();
		int m = pat.length();
		
		int i=0,j=0;
		int[] LPS = getLps(pat, m);

		while( i< ) {
			if(text.charAt(i)==pat.charAt(j)){
				i++; j++;
			}
			
			if(j==m){
				  System.out.println("Found pattern at index " + (i - j));
				  j=LPS[j-1];
			}
			else if(i < n&& pat.charAt(j) != text.charAt(i)) {
				
				if(j!=0)
					j=LPS[j-1];
				else
					i++;
				
			}
		}
	}

	public static int[] getLps(String pat, int m) {
		int[] LPS = new int[m];
		int i = 1;
		LPS[0] = 0;
		int len = 0;
		while (i < m) {
			if (pat.charAt(i) == pat.charAt(len)) {
				len++;
				LPS[i] = len;
				i++;
			} else {
				if (len > 0) {
					len = LPS[len - 1];
				} else {
					LPS[i] = len;
					i++;
				}
			}
		}
		
		return LPS;
	}
}

File: Other Algorithms/KMP/golang/kmp.go
//Implementation of Knuth Morris Pratt (KMP) algorithm for pattern searching

// Author: Saloni Rajeev Kumar
// Date: Oct 7, 2018

package main
import (
	"fmt"
	"bufio"
	"os"
)

var flag = 0

func calculate_lps(pat string, m int, lps []int){
  len := 0 // length of the previous longest suffix
  lps[0] = 0

  i:=1
  for i<m {
    if pat[i]==pat[len]{
      len++
      lps[i] = len
      i++
    } else {
      if len!=0{
        len = lps[len-1]
      } else {
        lps[i]=0
        i++
      }
    }
  }
}

func kmp_search(pat string, text string){
  n := len(text)-1 // to compensate for \n at the end
  m := len(pat)-1

  lps := make([]int, m)

  calculate_lps(pat, m, lps)

  i := 0
  j := 0

  for i<n {
    if pat[j]==text[i]{
      j++
      i++
    }

    if j==m {
      fmt.Println("Found pattern at address", i-j)
      j = lps[j-1]
      flag++
    } else if i<n && pat[j]!=text[i] {
      if j!=0{
        j=lps[j-1]
      } else {
        i++
      }
    }
  }
  if flag==0 {
    fmt.Println("Not found")
  }
}


func main(){
	reader := bufio.NewReader(os.Stdin)

  // Read text from user
  fmt.Println("Enter text: ")
  text, _ := reader.ReadString('\n')

  // Read pattern to be matched from user
  fmt.Println("Enter pattern: ")
  pat, _ := reader.ReadString('\n')

  kmp_search(pat, text)

}
